modules/core/src/main/java/org/apache/ignite/internal/processors/cache/ClusterCachesInfo.java

```diff
@@ -917,35 +920,39 @@ private boolean processStartNewCacheRequest(
         DynamicCacheChangeRequest req,
         String cacheName
     ) {
-        String conflictErr = checkCacheConflict(req.startCacheConfiguration());
+        CacheConfiguration<?, ?> ccfg = req.startCacheConfiguration();
 
-        if (conflictErr != null) {
-            U.warn(log, "Ignore cache start request. " + conflictErr);
+        IgniteCheckedException err = null;
 
-            IgniteCheckedException err = new IgniteCheckedException("Failed to start " +
-                "cache. " + conflictErr);
+        IgniteCacheSnapshotManager snapshotMgr = ctx.cache().context().snapshot();
 
-            if (persistedCfgs)
-                res.errs.add(err);
-            else
-                ctx.cache().completeCacheStartFuture(req, false, err);
+        String dataReg = ccfg.getDataRegionName();
 
-            return false;
+        if (ctx.cache().context().readOnlyMode() && !snapshotMgr.restoreOrRecoveryInProgress() &&
+            !VOLATILE_DATA_REGION_NAME.equals(dataReg)) {
+            err = new IgniteClusterReadOnlyException(
+                String.format(CLUSTER_READ_ONLY_MODE_ERROR_MSG_FORMAT, "start cache", ccfg.getGroupName(), cacheName)
+            );
         }
 
-        SchemaOperationException err = QueryUtils.checkQueryEntityConflicts(
-            req.startCacheConfiguration(), registeredCaches.values());
+        if (!validateStartNewCache(err, persistedCfgs, res, req))
+            return false;
 
-        if (err != null) {
-            if (persistedCfgs)
-                res.errs.add(err);
-            else
-                ctx.cache().completeCacheStartFuture(req, false, err);
+        String conflictErr = checkCacheConflict(req.startCacheConfiguration());
 
-            return false;
+        if (conflictErr != null) {
+            U.warn(log, "Ignore cache start request. " + conflictErr);
+
+            err = new IgniteCheckedException("Failed to start cache. " + conflictErr);
         }
 
-        CacheConfiguration<?, ?> ccfg = req.startCacheConfiguration();
+        if (!validateStartNewCache(err, persistedCfgs, res, req))
```

Comment:
validateStartNewCache is checked three times here. Is it intentional?

Reply:
yes, cause we have mutable **err**  passing state.